src/librustc/middle/ty/relate.rs

```diff
@@ -147,12 +238,15 @@ fn relate_substs<'a,'tcx:'a,R>(relation: &mut R,
     where R: TypeRelation<'a,'tcx>
 {
     let mut substs = Substs::empty();
+    let mut obligations = Vec::new();
 
     for &space in &ParamSpace::all() {
         let a_tps = a_subst.types.get_slice(space);
         let b_tps = b_subst.types.get_slice(space);
         let t_variances = variances.map(|v| v.types.get_slice(space));
-        let tps = try!(relate_type_params(relation, t_variances, a_tps, b_tps));
+        let RelateOk { value: tps, obligations: new_obligations } =
+            try!(relate_type_params(relation, t_variances, a_tps, b_tps));
+        obligations.extend(new_obligations);
```

Comment:
might be good to have a method like 

```
fn into_value(self, obligations_vec: &mut Vec<Obligation>) {
    obligations_vec.extend(self.obligations);
    self.value
}
```

then you could just write `let tps = try!(...).into_value(&mut obligations);`